Applications/Embedding/jni/main.cpp

```diff
@@ -0,0 +1,282 @@
+      outVec[0][count * feature_size + j] = o[j];
+    outLabel[0][count] = l[0];
+
+    count++;
+  }
+
+  dataFile.close();
+  *last = false;
+  train_count += batch_size;
+  return 0;
+}
+
+/**
+ * @brief     create NN
+ *            back propagation of NN
+ * @param[in]  arg 1 : train / inference
+ * @param[in]  arg 2 : configuration file path
+ * @param[in]  arg 3 : resource path (dataset.txt or testset.txt)
+ */
+int main(int argc, char *argv[]) {
+  if (argc < 4) {
+    std::cout << "./Embedding train (| inference) Config.ini data.txt\n";
+    exit(1);
+  }
+
+  const std::vector<std::string> args(argv + 1, argv + argc);
+  std::string config = args[1];
+  data_file = args[2];
+
+  if (!args[0].compare("train"))
+    training = true;
+
+  srand(time(NULL));
+
+  std::shared_ptr<ml::train::Dataset> dataset;
+  try {
+    dataset = createDataset(ml::train::DatasetType::GENERATOR);
+    dataset->setGeneratorFunc(ml::train::DatasetDataType::DATA_TRAIN,
+                              getBatch_train);
+    dataset->setGeneratorFunc(ml::train::DatasetDataType::DATA_VAL,
+                              getBatch_train);
+  } catch (std::exception &e) {
+    std::cerr << "Error creating dataset" << e.what() << std::endl;
+    return 1;
+  }
+
+  /**
+   * @brief     Create NN
+   */
+  std::vector<std::vector<float>> inputVector, outputVector;
+  nntrainer::NeuralNetwork NN;
+  std::shared_ptr<nntrainer::Layer> layer;
+  std::shared_ptr<nntrainer::Layer> layer_fc;
+  std::string name = "embedding";
+  std::string fc_name = "outputlayer";
+  nntrainer::Tensor weight;
+  nntrainer::Tensor weight_fc;
+  /**
+   * @brief     Initialize NN with configuration file path
+   */
+
+  try {
+    NN.loadFromConfig(config);
+    NN.compile();
+    NN.initialize();
+    NN.readModel();
+
+    nntrainer::Tensor weight_in(1, 1, 15, 8);
+    loadFile("embedding_weight_input.in", weight_in);
+
+    weight_in.print(std::cout);
+
+    NN.getLayer(name.c_str(), &layer);
+    NN.getLayer(fc_name.c_str(), &layer_fc);
+
+    weight = layer->getWeights()[0].getVariable();
+    weight.print(std::cout);
+
+  } catch (...) {
+    std::cerr << "Error during init" << std::endl;
+    return 0;
+  }
+
+  if (training) {
+    NN.setDataset(dataset);
+    try {
+      NN.train();
+    } catch (...) {
+      std::cerr << "Error during train" << std::endl;
+      return 0;
+    }
+
+    // std::shared_ptr<nntrainer::Layer> layer;
+    // std::string name = "embedding";
+    // NN.getLayer(name.c_str(), &layer);
+
+    weight = layer->getWeights()[0].getVariable();
+    weight.print(std::cout);
+
+    nntrainer::Tensor golden(1, 1, 15, 8);
+    loadFile("embedding_weight_golden.out", golden);
+
+    golden.print(std::cout);
+
+    nntrainer::Tensor weight_out_fc(1, 1, 32, 1);
+    loadFile("fc_weight_golden.out", weight_out_fc);
+    weight_out_fc.print(std::cout);
+
+    weight_fc = layer_fc->getWeights()[0].getVariable();
+    weight_fc.print(std::cout);
```

Comment:
let's add dependency on gtest and verify values than just printing, as there is no unittest for training embedding layer, or add some unittests for embedding layer training as well.

Reply:
later pr. I guess.

Reply:
alright